File: src/clients/sbig_client.py
#!/usr/bin/env python

#    ICE Telescope ROS Package: SBIG CCD Client
#    Copyright (C) 2015 Biel Artigues Aguilo
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

import sys
import rospy
from ice_telescope.srv import *

def usage_general():
	rospy.loginfo("Usage: sbig_client action [action arguments]. Action: capture; settemp; gettemp; getcapstatus; reconnect")

def usage_capture():
	rospy.loginfo("Usage capture: sbig_client capture filePath fileType imgCount imgType expTime readoutMode top left width height fastReadout dualReadoutChannel")
	rospy.loginfo("Example: sbig_client capture /observations/img/raw/ FITS 1 LF 0.01 1x1 0 0 0 0 1 1")
	rospy.loginfo("Example: sbig_client capture /observations/img/raw/ SBIG 1 LF 0.01 1x1 0 0 0 0 0 0")

def usage_settemp():
	rospy.loginfo("Usage settemp: sbig_client settemp enable temperature")
	rospy.loginfo("Example: sbig_client settemp 1 10.0")

def sbig_action_client(sbig_action, file_path, fits_file, img_count, lf_img, exp_time, readout_mode, top, left, width, height, fast_readout, dual_readout_channel, temperature, temp_enable):
    rospy.wait_for_service('sbig_action')
    try:
        srv = rospy.ServiceProxy('sbig_action', sbig)
        if sbig_action == "capture":
        	rospy.loginfo("Starting CCD exposure")
        resp = srv(sbig_action, file_path, fits_file, img_count, lf_img, exp_time, readout_mode, top, left, width, height, fast_readout, dual_readout_channel, temperature, temp_enable)
        return resp
    except rospy.ServiceException, e:
        rospy.logerr("Service call failed: %s", e)


if __name__ == "__main__":
	rospy.init_node('sbig_client')
	if len(sys.argv) < 2:
		usage_general()
		sys.exit(1)

	sbig_action = str(sys.argv[1])

	if sbig_action == "capture":
		if len(sys.argv) != 14:
			usage_capture()
			sys.exit(1)

		file_path = str(sys.argv[2])
		if not file_path.endswith("/"):
			file_path = file_path + '/'

		fits_file = False
		if str(sys.argv[3]) == "FITS":
			fits_file = True

		img_count = int(sys.argv[4])

		lf_img = False
		if str(sys.argv[5]) == "LF":
			lf_img = True

		exp_time = float(sys.argv[6])

		readout_mode = 0
		if str(sys.argv[7]) == "2x2":
			readout_mode = 1
		elif str(sys.argv[7]) == "3x3":
			readout_mode = 2

		top = int(sys.argv[8])
		left = int(sys.argv[9])
		width = int(sys.argv[10])
		height = int(sys.argv[11])

		fast_readout = False
		if str(sys.argv[12]) == "1":
			fast_readout = True

		dual_readout_channel = False
		if str(sys.argv[13]) == "1":
			dual_readout_channel = True

		temperature = None
		temp_enable = None

	elif sbig_action == "settemp":
		if len(sys.argv) != 4:
			usage_settemp()
			sys.exit(1)

		file_path = None
		fits_file = None
		img_count = None
		lf_img = None
		exp_time = None
		readout_mode = None
		top = None
		left = None
		width = None
		height = None
		fast_readout = None
		dual_readout_channel = None
		temperature = float(sys.argv[3])
		temp_enable = False
		if str(sys.argv[2]) == "1":
			temp_enable = True

	else:
		file_path = None
		fits_file = None
		img_count = None
		lf_img = None
		exp_time = None
		readout_mode = None
		top = None
		left = None
		width = None
		height = None
		fast_readout = None
		dual_readout_channel = None
		temperature = None
		temp_enable = None


	resp = sbig_action_client(sbig_action, file_path, fits_file, img_count, lf_img, exp_time, readout_mode, top, left, width, height, fast_readout, dual_readout_channel, temperature, temp_enable)
	if resp.sbig_error:
		rospy.logerr("%s", resp.sbig_response)
		sys.exit(1)
	else:
		rospy.loginfo("%s", resp.sbig_response)
		sys.exit(0)

File: src/clients/sbig_client.cpp
/*
    ICE Telescope ROS Package: SBIG CCD Client
    Copyright (C) IEEC 2020
    Author: Biel Artigues Aguilo

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "ros/ros.h"
#include "ice_telescope/sbig.h"
#include <string.h>

void usage_general()
{
  ROS_INFO("Usage: sbig_client action [action arguments]. Action: capture; settemp; gettemp; getcapstatus; reconnect");
}

void usage_capture()
{
  ROS_INFO("Usage capture: sbig_client capture filePath fileType imgCount imgType expTime readoutMode top left width height fastReadout dualReadoutChannel");
  ROS_INFO("Example: sbig_client capture /observations/img/raw/ FITS 1 LF 0.01 1x1 0 0 0 0 1 1");
  ROS_INFO("Example: sbig_client capture /observations/img/raw/ SBIG 1 LF 0.01 1x1 0 0 0 0 0 0");
}

void usage_settemp()
{
  ROS_INFO("Usage settemp: sbig_client settemp enable temperature");
  ROS_INFO("Example: sbig_client settemp 1 10.0");
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "sbig_client");
  if(argc < 2)
  {
    usage_general();
    return 1;
  }

  ros::NodeHandle n;
  ros::ServiceClient client = n.serviceClient<ice_telescope::sbig>("sbig_action");
  ice_telescope::sbig srv;

  // Action
  srv.request.sbig_action = argv[1];

  if(srv.request.sbig_action == "capture")
  {
    if(argc != 14)
    {
      usage_capture();
      return 1;
    }

    // File path
    srv.request.file_path = argv[2];
    int len = srv.request.file_path.length();
    if(srv.request.file_path[len - 1] != '/')
    {
      srv.request.file_path.append("/");
    }

    // File type
    srv.request.fits_file = false;
    if(strcmp(argv[3], "FITS") == 0)
    {
      srv.request.fits_file = true;
    }

    // Image count
    srv.request.img_count = atoi(argv[4]);

    // Image type. Suppose DF
    srv.request.lf_img = false;
    if(strcmp(argv[5], "LF") == 0)
    {
      srv.request.lf_img = true;
    }

    // Exposure time
    srv.request.exp_time = atof(argv[6]);

    // Readout mode. Suppose 1x1
    srv.request.readout_mode = 0;
    if(strcmp(argv[7], "2x2") == 0)
    {
      srv.request.readout_mode = 1;
    }
    else if(strcmp(argv[7], "3x3") == 0)
    {
      srv.request.readout_mode = 2;
    }

    // Image size: if all params, ie. top, left, width and height are zero,
    // the full size of the CCD image is used.
    srv.request.top = atoi(argv[8]);
    srv.request.left = atoi(argv[9]);
    srv.request.width = atoi(argv[10]);
    srv.request.height = atoi(argv[11]);

    // Fast readout
    srv.request.fast_readout = false;
    if(strcmp(argv[12], "1") == 0)
    {
      srv.request.fast_readout = true;
    }

    // Dual channel mode
    srv.request.dual_readout_channel = false;
    if(strcmp(argv[13], "1") == 0)
    {
      srv.request.dual_readout_channel = true;
    }

    ROS_INFO("Starting CCD exposure");
  }
  else if(srv.request.sbig_action == "settemp")
  {
    if(argc != 4)
    {
      usage_settemp();
      return 1;
    }

    srv.request.temperature = atof(argv[3]);
    srv.request.temp_enable = false;
    if(strcmp(argv[2], "1") == 0)
    {
      srv.request.temp_enable = true;
    }
  }

  if(client.call(srv))
  {
    if(srv.response.sbig_error)
    {
      ROS_ERROR(srv.response.sbig_response.c_str());
      return 1;
    }
    else
    {
      ROS_INFO(srv.response.sbig_response.c_str());
      return 0;
    }
  }
  else
  {
    ROS_ERROR("Failed to call ccd service");
    return 1;
  }
}
